src/blend.cpp

```diff
@@ -315,81 +312,106 @@ void Work_Blend(uv_work_t* req)
+                continue;
+            }
+            image->width = layer_width;
+            image->height = layer_height;
+            image->im_raw_ptr = &image->im_obj->get()->get<mapnik::image_rgba8>();
+        } else {
+            std::unique_ptr<mapnik::image_reader> image_reader;
+            try {
+                image_reader = std::unique_ptr<mapnik::image_reader>(mapnik::get_image_reader(image->data, image->dataLength));
+            } catch (std::exception const& ex) {
+                baton->message = ex.what();
+                return;
+            }
 
-        unsigned layer_width = image_reader->width();
-        unsigned layer_height = image_reader->height();
-        // Error out on invalid images.
-        if (layer_width == 0 || layer_height == 0) {
-            // No idea how to create a zero height or width image
-            // so removing from coverage, because I am fairly certain
-            // it is not possible in almost every image format.
-            /* LCOV_EXCL_START */
-            baton->message = "zero width/height image encountered";
-            return;
-            /* LCOV_EXCL_STOP */
-        }
+            if (!image_reader || !image_reader.get()) {
+                // Not quite sure anymore how the pointer would not be returned
+                // from the reader and can't find a way to make this fail.
+                // So removing from coverage
+                /* LCOV_EXCL_START */
+                baton->message = "Unknown image format";
+                return;
+                /* LCOV_EXCL_STOP */
+            }
 
-        int visibleWidth = (int)layer_width + image->x;
-        int visibleHeight = (int)layer_height + image->y;
-        // The first image that is in the viewport sets the width/height, if not user supplied.
-        if (baton->width <= 0) baton->width = std::max(0, visibleWidth);
-        if (baton->height <= 0) baton->height = std::max(0, visibleHeight);
+            unsigned layer_width = image_reader->width();
+            unsigned layer_height = image_reader->height();
+            // Error out on invalid images.
+            if (layer_width == 0 || layer_height == 0) {
+                // No idea how to create a zero height or width image
+                // so removing from coverage, because I am fairly certain
+                // it is not possible in almost every image format.
+                /* LCOV_EXCL_START */
+                baton->message = "zero width/height image encountered";
+                return;
+                /* LCOV_EXCL_STOP */
+            }
 
-        // Skip images that are outside of the viewport.
-        if (visibleWidth <= 0 || visibleHeight <= 0 || image->x >= baton->width || image->y >= baton->height) {
-            // Remove this layer from the list of layers we consider blending.
-            continue;
-        }
+            int visibleWidth = (int)layer_width + image->x;
```

Comment:
nitpick: any reason not to use `std::int32_t` here?

Reply:
Probably should be int64_t or int32_t, but this is a base type that is used in lots of places in the current code, not going to redo it all yet because not trying to modify the code too much  for this change. At some point we still need to just do a sweeping review of current node mapnik design and update all the slight issues in the current setup.